docs/proposals/20200615-external-remediation.md

```diff
@@ -0,0 +1,299 @@
+
+- Facilitate rapid experimentation by creating the ability to define customized remediation flows outside of the Machine Health Check and CAPI codebase.
+
+### Non-Goals/Future Work
+
+- Provide a [unified remediation flow](https://docs.google.com/document/d/1klvIR5wKbcCrl2vLPIbbWpmelGTDM-T0Qb4xw36P3zY/edit) as part of the MachineHealthCheck controller. See Alternative 1
+- Add power primitives to the Machine API.  While useful, it is unnecessary to the work proposed here and may already be covered by one or more of the following issues:
+    - Add Ability to Reboot Hosts
+    - Supporting power off/on
+    - Add machine/infra feature support fields
+- Alter the default deletion flow proposed in [cluster-api PR#3056](https://github.com/kubernetes-sigs/cluster-api/pull/3056)
+- Provide a finalizer-like pre-hook mechanism for allowing customization or blocking of the remediation process prior to power-cycling a machine or deleting it. This concept may already be covered as part of a separate enhancement
+- Create an escalation path from failed external remediation attempts to machine deletion.
+- Implement a formal mechanism for triggering recovery of workload and volumes by the k8s scheduler
+
+## Proposal
+
+We introduce a generic mechanism for supporting externally provided custom remediation strategies.
+
+We propose modifying the MachineHealthCheck CRD to support a remediationTemplate, an ObjectReference to a provider-specific template CRD. 
+
+If no value for remediationTemplate is defined for the MachineHealthCheck CR, the existing condition-based deletion flow is preserved.
+
+If a value for remediationTemplate is supplied and the Machine enters an unhealthy state, the template will be instantiated using existing CAPI functionality, with the same name and namespace as the target Machine, and the remediation flow passed to an External Remediation Controller (ERC) watching for that CR.  
+
+No further action (deletion or applying conditions) will be taken by the MachineHealthCheck controller until the Node becomes healthy, when it will locate and delete the instantiated MachineRemediation CR.
+
+The ERC may wish to register a finalizer on its CR to ensure it has an opportunity to perform any additional cleanup in the case that the unhealthy state was transient and the Node returned to a healthy state prior to the completion of the full custom ERC flow.
+
+
+### User Stories
+
+#### Story 1
+As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can recover from transient errors faster and begin application recovery sooner.
+#### Story 2
+As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can detect hardware issues faster.
+#### Story 3
+As an admin of a hardware based cluster, I would like the system to keep attempting to power-cycle unhealthy nodes, so that they are automatically added back to the cluster when I fix the underlying problem.
```

Comment:
> Can you provide an example of when this would be intended behavior? 

routing issues for example that are external to the server itself but prevents it to reach the cluster network

> why would we want to continually restart until a technician can fix the server rather than trying to provision on a different server?

That's a good point, and I think that in one of the discussion it was suggested to have some kind of max reboot attempts, but I think this is up to the external remediation controller to decide